framework/src/main/java/org/checkerframework/framework/type/DefaultInferredTypesApplier.java

```diff
@@ -90,6 +91,12 @@ private void apply(
                 }
             } else if ((omitSubtypingCheck || hierarchy.isSubtype(inferred, primary))) {
                 type.replaceAnnotation(inferred);
+                if (type.getKind() == TypeKind.INTERSECTION) {
```

Comment:
It's not your fault, but the code is confusing when juxtaposed with the code just above it (lines 84-91).
I wonder if that can be refactored to emphasize the commonality, maybe in a separate pull request.
For example, that might clarify why the added code is not necessary when primary is null.